### PR TITLE
Allow vaulted string to be used as actual strings (#73054)

### LINKED ISSUE #73054: urlencode filter doesn't accept encrypted variable
##### SUMMARY

When trying to pass a encrypted variable to the `urlencode` filter ansible 2.10 returns an error.

This does not affect ansible 2.9.16. 

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
filter

##### ANSIBLE VERSION
```
ansible 2.11.0.dev0
  config file = None
  configured module search path = ['/home/username/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/username/gits/ansible-debug/.venv/ansible-devel/lib/python3.8/site-packages/ansible
  ansible collection location = /home/username/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/username/gits/ansible-debug/.venv/ansible-devel/bin/ansible
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
  jinja version = 2.11.2
  libyaml = False
```

```
ansible 2.10.4
  config file = None
  configured module search path = ['/home/username/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/username/gits/ansible-debug/.venv/ansible-2.10/lib/python3.8/site-packages/ansible
  executable location = /home/username/gits/ansible-debug/.venv/ansible-2.10/bin/ansible
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]

```

##### CONFIGURATION

Default Configuration

##### OS / ENVIRONMENT

Ubuntu 20.04.1 LTS

##### STEPS TO REPRODUCE
Use the sample playbook below, the vault password is `test`.

```yaml
---
- hosts: localhost
  vars:
    vaulted_var: !vault |
        $ANSIBLE_VAULT;1.1;AES256
        30353233336165306463306539316563386232666633346430626465393435663139396134303638
        6539303730373265346461386561333333383139316131650a303833363362643564373030383637
        37303732373564636666313431316664363262363836346231373334333838343464343634356165
        3039316538316234380a396534623733343633623831393736333533333438393034353835633764
        63393061633561306230633964343531333537393439363035356335666238363432

  tasks:
  - name: vaulted_var | urlencode
    debug:
      msg: "{{ vaulted_var | urlencode }}"
    ignore_errors: true
  - name: 'workaround: "{}".format(vaulted_var) | urlencode'
    debug:
      msg: '{{ "{}".format(vaulted_var) | urlencode }}'

```

##### EXPECTED RESULTS

The `vaulted_var` should be passed to `urlencode` and return the value of `vaulted_var` with url encoding.

vaulted_var: `this is a test!@#$%^&*()[]{}~/.`

expected result: `this%20is%20a%20test%21%40%23%24%25%5E%26%2A%28%29%5B%5D%7B%7D~/.`

##### ACTUAL RESULTS

```
TASK [vaulted_var | urlencode] ***********************************************************************************************************************************************************
task path: /home/username/gits/ansible-debug/sample.play:12
Found a vault_id (default) in the vaulttext
We have a secret associated with vault id (default), will try to use to decrypt None                                                                                                      
Trying to use vault secret=(FileVaultSecret(filename='/home/username/gits/ansible-debug/.vault_pass')) id=default to decrypt None
Trying secret FileVaultSecret(filename='/home/username/gits/ansible-debug/.vault_pass') for vault_id=default
Decrypt successful with secret=FileVaultSecret(filename='/home/username/gits/ansible-debug/.vault_pass') and vault_id=default
fatal: [localhost]: FAILED! => {                                                             
    "msg": "the field 'args' has an invalid value ({'msg': '{{ vaulted_var | urlencode }}'}), and could not be converted to an dict.The error was: not enough values to unpack (expected 2
, got 1)\n\nThe error appears to be in '/home/username/gits/ansible-debug/sample.play': line 12, column 5, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe o
ffending line appears to be:\n\n  tasks:\n  - name: vaulted_var | urlencode\n    ^ here\n"
}
...ignoring
```

### PR DESCRIPTION
**Pull request by eqrx**
##### SUMMARY
Multiple places in the code base expect actual `str` types and
refuse to work with objects that can be treated as strings.
This commit lets `AnsibleVaultEncryptedUnicode` deceive those
places by faking to be a `str`.

Fixes #73054

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`AnsibleVaultEncryptedUnicode`

### CODE REVIEW

File: lib/ansible/parsing/yaml/objects.py [hunk 1]
```diff
@@ -113,6 +113,9 @@ def __init__(self, ciphertext):
         self.vault = None
         self._ciphertext = to_bytes(ciphertext)
 
+    __class__ = text_type
+
```
Comment (bcoca) [inline lib/ansible/parsing/yaml/objects.py]: shouldn't we just inherit from AnsibleUnicode instead of AnsibleBaseYAMLObject

Comment (webknjaz) [reply lib/ansible/parsing/yaml/objects.py:117]: Yeah, this looks dangerous. But also inheriting from string is rather problematic. I recall attempting this a year ago in our migration script and it was rather painful.

Comment (eqrx) [reply lib/ansible/parsing/yaml/objects.py]: that is smarter :) agreed

**Comment by eqrx** [comment]
@bcoca looking at the failed test it seems that the change indeed fixes the direct bug, but lets other tests fail because the value gets serialized to another representation. What do you think would be the best approach? Would multi inheritance solve this problem without side effects, should I adapt the tests (sounds like a bad idea) or should I dig down more into the problem and try to find another solution (will do that until you answer :) )? Thank

**Comment by bcoca** [comment]
` serialized to another representation` this should always be the case with vaults .. its basically how it 'unvaults'

**Comment by eqrx** [comment]
(fyi, I working on this. Just taking longer since I need to take a closer look at the testing system)

File: lib/ansible/module_utils/common/json.py [hunk 2]
```diff
@@ -49,6 +49,15 @@ def __init__(self, preprocess_unsafe=False, vault_to_text=False, **kwargs):
...
+            # vault object
+            if self._vault_to_text:
+                o = to_text(o, errors='surrogate_or_strict')
+            else:
+                o = {'__ansible_vault': to_text(o._ciphertext, errors='surrogate_or_strict', nonstring='strict')}
+        return json.JSONEncoder.encode(self, o)
+
     # NOTE: ALWAYS inform AWS/Tower when new items get added as they consume them downstream via a callback
     def default(self, o):
         if getattr(o, '__ENCRYPTED__', False):
```
Comment (bcoca) [inline lib/ansible/module_utils/common/json.py:63]: do we still need this entry?

Comment (eqrx) [reply lib/ansible/module_utils/common/json.py:63]: well, if I leave this entry out the json encoder just encodes it as a string without consulting the `default` method, which breaks everything

Comment (bcoca) [reply lib/ansible/module_utils/common/json.py:63]: i mean the entry in the default that check encrypted, now that it is in encode, it should never get there.

Comment (eqrx) [reply lib/ansible/module_utils/common/json.py:63]: oh sorry, jup

**Comment by mkrizek** [comment]
!waffling core_review

**Comment by eqrx** [comment]
My solution does not work, closing

**Pull request closed without merging**